src/runtime/BackingStore.cpp

```diff
@@ -24,83 +24,122 @@
+    static MAY_THREAD_LOCAL GC_descr descr;
+    if (!typeInited) {
+        // only mark m_deleterData
+        GC_word obj_bitmap[GC_BITMAP_SIZE(NonSharedBackingStore)] = { 0 };
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(NonSharedBackingStore, m_deleterData));
+        descr = GC_make_descriptor(obj_bitmap, GC_WORD_LEN(NonSharedBackingStore));
+        typeInited = true;
+    }
+    return GC_MALLOC_EXPLICITLY_TYPED(size, descr);
 }
 
-void BackingStore::reallocate(size_t newByteLength)
+void NonSharedBackingStore::reallocate(size_t newByteLength)
 {
-    // Shared Data Block should not be reallocated
-    ASSERT(!m_isShared);
-
     if (m_byteLength == newByteLength) {
         return;
     }
 
-    if (m_isAllocatedByPlatformAllocator) {
+    if (m_isAllocatedByPlatform) {
         m_data = Global::platform()->onReallocArrayBufferObjectDataBuffer(m_data, m_byteLength, newByteLength);
         m_byteLength = newByteLength;
     } else {
+        // Note) even if BackingStore was previously allocated by other allocator,
+        // newly reallocate it with default Platform allocator
         m_deleter(m_data, m_byteLength, m_deleterData);
         m_data = Global::platform()->onMallocArrayBufferObjectDataBuffer(newByteLength);
-        m_deleter = defaultPlatformBackingStoreDeleter;
+        m_deleter = backingStorePlatformDeleter;
         m_deleterData = nullptr;
         m_byteLength = newByteLength;
-        m_isAllocatedByPlatformAllocator = true;
+        m_isAllocatedByPlatform = true;
+    }
+}
+
+#if defined(ENABLE_THREADING)
+BackingStore* BackingStore::createDefaultSharedBackingStore(size_t byteLength)
+{
+    SharedDataBlockInfo* sharedInfo = new SharedDataBlockInfo(
+        Global::platform()->onMallocArrayBufferObjectDataBuffer(byteLength),
+        byteLength);
+    return new SharedBackingStore(sharedInfo);
+}
+
+BackingStore* BackingStore::createSharedBackingStore(SharedDataBlockInfo* sharedInfo)
+{
+    ASSERT(sharedInfo->hasValidReference());
+    return new SharedBackingStore(sharedInfo);
+}
+
+void SharedDataBlockInfo::deref()
+{
+    ASSERT(hasValidReference());
+
+    auto oldValue = m_refCount.fetch_sub(1);
+    if (oldValue == 1) {
+        Global::platform()->onFreeArrayBufferObjectDataBuffer(m_data, m_byteLength);
+        m_data = nullptr;
+        m_byteLength = 0;
+
+        delete this;
+        return;
     }
 }
 
+SharedBackingStore::SharedBackingStore(SharedDataBlockInfo* sharedInfo)
+    : m_sharedDataBlockInfo(sharedInfo)
+{
+    // increase reference count
+    sharedInfo->ref();
+
+    GC_REGISTER_FINALIZER_NO_ORDER(this, [](void* obj, void*) {
+        SharedBackingStore* self = (SharedBackingStore*)obj;
+        self->sharedDataBlockInfo()->deref();
+    },
+                                   nullptr, nullptr, nullptr);
+}
+
+void* SharedBackingStore::operator new(size_t size)
+{
+    // SharedBackingStore does not have any GC member
+    return GC_MALLOC_ATOMIC(size);
```

Comment:
IMO we should not use GC_MALLOC_ATOMIC
we can just use malloc and free
because we cannot support cross-thread gc, malloc free
we should isolate gc allocated memory per thread

Reply:
I cannot fully understand your comment (correct me if I'm wrong)
`SharedBackingStore` is created per thread not shared by several threads.
`SharedDataBlockInfo` is the only shared structure and it is allocated by basic `new` operator.

So, when a SharedArrayBuffer is shared between workers(threads),
actually each SharedArrayBuffer and its SharedBackingStore is newly generated for each thread and the origin SharedDataBlockInfo is shared by all SharedBackingStore having that one SharedBackingInfo in itself.

Reply:
the name of `SharedBackingStore` is confusing I think(cause it can be used with SharedArrayBuffer)
I understand that user should not share `SharedBackingStore`
How about to add simple comment to SharedBackingStore in public header?
eg) u can only share SharedDataBlockInfo across threads.

Reply:
I update some comment for it. Thanks!